scalyr_agent/log_processing.py

```diff
@@ -88,45 +88,150 @@
-    timestamp = rfc3339_to_nanoseconds_since_epoch(line[:index])
-    if timestamp is None:
-        return None, None, None, None
+    @staticmethod
+    def __parse_cri_log(line):
+        """
+        Parse a log line that uses the K8S Container Runtime Interface (CRI) format
+        https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/kubelet-cri-logging.md
+        """
+        parts = line.split(" ", maxsplit=3)
+        # The 4 parts are the timestamp, stream, tags, and message. The 4th part can contain spaces.
+        if len(parts) < 3:
+            raise CRIParseError("Expected 3 or 4 records (timestamp, stream, tags, optional message), got %d" % len(parts), line)
 
-    # skip over the ' ' delimiter and parse the stream type
-    line = line[index + 1 :]
-    index = line.find(" ")
-    if index < 0:
-        return None, None, None, None
+        # parse the timestamp
+        timestamp = rfc3339_to_nanoseconds_since_epoch(parts[0])
+        if timestamp is None:
+            raise CRIParseError("Could not parse timestamp", line)
+        raw_timestamp = timestamp
 
-    stream = line[:index]
-    if stream != "stdout" and stream != "stderr":
-        return None, None, None, None
+        # parse the stream type
+        stream = parts[1]
+        if stream != "stdout" and stream != "stderr":
+            raise CRIParseError("Unknown stream type", line)
 
-    # skip over the ' ' delimiter and parse the tags
-    line = line[index + 1 :]
-    index = line.find(" ")
-    if index < 0:
-        return None, None, None, None
+        tags = parts[2].split(":")
+        line = parts[3] if len(parts) == 4 else ""
 
-    tags = line[:index]
+        return timestamp, stream, tags, line, raw_timestamp
 
-    # skip over the ' ' delimiter and get the log line
-    line = line[index + 1 :]
+class CRILogLineBuffer(object):
+    def __init__(self, first_line_time):
+        self.__stream = None
+        self.__lines = []
+        self.__timestamp = None
+        self.__raw_timestamp = None
+        self.__is_partial = True
+        self.__first_line_time = first_line_time
 
-    return timestamp, stream, tags, line
+    def append(self, log_line):
+        if not self.__lines:
+            self.__timestamp = log_line.timestamp
+            self.__raw_timestamp = log_line.raw_timestamp
+            self.__stream = log_line.stream
 
+        self.__lines.append(log_line.message)
+
+        if log_line.is_full:
+            self.__is_partial = False
+
+    @property
+    def stream(self):
+        return self.__stream
+
+    @property
+    def first_line_time(self):
+        return self.__first_line_time
+
+    @property
+    def is_partial(self):
+        return self.__is_partial
+
+    def payload_length(self):
+        return sum(map(len, self.__lines))
+
+    def build_cri_result(self, include_raw_timestamp_field):
+        def strip_new_line(s):
+            if s and s[-1] == "\n":
```

Comment:
Could the lines also contains carriage retention character `\r` (e.g. on Window host)? And if so, do we need / want to strip it?

Reply:
Yes, I will add it there.